src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,149 @@
+            return;
+        }
+        checkIndex(index);
+        Node<T> currentNode = node(index);
+        Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+        Node<T> previousNode = currentNode.prev;
+        if (previousNode == null) {
+            head = newNode;
+        } else {
+            previousNode.next = newNode;
+        }
+        currentNode.prev = newNode;
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T element : list) {
+            add(element);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return node(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nodeToChange = node(index);
+        T oldItem = nodeToChange.item;
+        nodeToChange.item = value;
+        return oldItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> oldNode = node(index);
+        return unlink(oldNode);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if ((currentNode.item == object)
+                    || (object != null && object.equals(currentNode.item))) {
+                remove(i);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("No such index in data");
+        }
+    }
+
+    private Node<T> node(int index) {
+        Node<T> currentNode;
+        if (index <= size / 2) {
+            currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+            return currentNode;
+        }
+        currentNode = tail;
+        for (int i = size - 1; i > index; i--) {
+            currentNode = currentNode.prev;
+        }
+        return currentNode;
+    }
+
+    private T unlink(Node<T> oldNode) {
+        Node<T> nextNode = oldNode.next;
+        Node<T> previousNode = oldNode.prev;
+        if (size == 1) {
+            head = null;
+            tail = null;
+        } else if (oldNode == head) {
+            nextNode.prev = null;
+            head = nextNode;
+        } else if (oldNode == tail) {
+            previousNode.next = null;
+            tail = previousNode;
+        } else {
+            nextNode.prev = previousNode;
+            previousNode.next = nextNode;
+        }
```

Comment:
let's try to simplify this a bit

first check if `nextNode` is null - write `if-else` construction

then check if `previousNode` is null - write another `if-else` construction

let's also set node's value to null in the end